### PR TITLE
IGNITE-13349: Tcp discovery statistics migrated to the new metrics framework.

### PR DESCRIPTION
**Pull request by nizhikov**
Thank you for submitting the pull request to the Apache Ignite.

In order to streamline the review of the contribution 
we ask you to ensure the following steps have been taken:

### The Contribution Checklist
- [ ] There is a single JIRA ticket related to the pull request. 
- [ ] The web-link to the pull request is attached to the JIRA ticket.
- [ ] The JIRA ticket has the _Patch Available_ state.
- [ ] The pull request body describes changes that have been made. 
The description explains _WHAT_ and _WHY_ was made instead of _HOW_.
- [ ] The pull request title is treated as the final commit message. 
The following pattern must be used: `IGNITE-XXXX Change summary` where `XXXX` - number of JIRA issue.
- [ ] A reviewer has been mentioned through the JIRA comments 
(see [the Maintainers list](https://cwiki.apache.org/confluence/display/IGNITE/How+to+Contribute#HowtoContribute-ReviewProcessandMaintainers)) 
- [ ] The pull request has been checked by the Teamcity Bot and 
the `green visa` attached to the JIRA ticket (see [TC.Bot: Check PR](https://mtcga.gridgain.com/prs.html))

### Notes
- [How to Contribute](https://cwiki.apache.org/confluence/display/IGNITE/How+to+Contribute)
- [Coding abbreviation rules](https://cwiki.apache.org/confluence/display/IGNITE/Abbreviation+Rules)
- [Coding Guidelines](https://cwiki.apache.org/confluence/display/IGNITE/Coding+Guidelines)
- [Apache Ignite Teamcity Bot](https://cwiki.apache.org/confluence/display/IGNITE/Apache+Ignite+Teamcity+Bot)

If you need any help, please email dev@ignite.apache.org or ask anу advice on http://asf.slack.com _#ignite_ channel.

### CODE REVIEW

**Comment by nizhikov** [comment]
Hello, @ingvard can you, please, take a look at my changes?

File: modules/core/src/main/java/org/apache/ignite/spi/discovery/tcp/internal/TcpDiscoveryStatistics.java [hunk 1]
```diff
@@ -287,23 +158,18 @@ public synchronized void onMessageProcessingStarted(TcpDiscoveryAbstractMessage
     public synchronized void onMessageProcessingFinished(TcpDiscoveryAbstractMessage msg) {
         assert msg != null;
 
-        Long startTs = msgsProcStartTs.get(msg.id());
+        Long startTs = msgsProcStartTs.remove(msg.id());
 
-        if (startTs != null) {
+        if (startTs == null) {
```
Comment (ingvard) [inline modules/core/src/main/java/org/apache/ignite/spi/discovery/tcp/internal/TcpDiscoveryStatistics.java:163]: Is it mistake?  long duration = U.currentTimeMillis() - startTs;  startTs is always null.

File: modules/core/src/main/java/org/apache/ignite/spi/discovery/tcp/TcpDiscoverySpi.java [hunk 2]
```diff
@@ -2115,6 +2115,20 @@ protected void onExchange(DiscoveryDataPacket dataPacket, ClassLoader clsLdr) {
     @Override public void spiStart(@Nullable String igniteInstanceName) throws IgniteSpiException {
         initializeImpl();
 
+        MetricRegistry discoReg =
+            ((IgniteEx)ignite()).context().metric().registry(metricName("io", "discovery"));
```
[the remark this answers is not in the document]
Comment (nizhikov) [reply modules/core/src/main/java/org/apache/ignite/spi/discovery/tcp/TcpDiscoverySpi.java]: Fixed.

File: modules/core/src/main/java/org/apache/ignite/spi/discovery/tcp/internal/TcpDiscoveryStatistics.java (hunk 1 above)
Comment (nizhikov) [reply modules/core/src/main/java/org/apache/ignite/spi/discovery/tcp/internal/TcpDiscoveryStatistics.java]: Yes, my fault.

Thanks for noticing.

File: modules/core/src/main/java/org/apache/ignite/spi/discovery/tcp/TcpDiscoverySpi.java [hunk 3]
```diff
@@ -1358,15 +1363,6 @@ public long getPendingMessagesRegistered() {
         return stats.pendingMessagesRegistered();
     }
 
-    /**
-     * Gets pending messages discarded count.
-     *
-     * @return Pending messages registered count.
-     */
-    public long getPendingMessagesDiscarded() {
```
Comment (NSAmelchev) [inline modules/core/src/main/java/org/apache/ignite/spi/discovery/tcp/TcpDiscoverySpi.java:1366]: Why the public API was changed?

File: modules/core/src/main/java/org/apache/ignite/spi/discovery/tcp/TcpDiscoverySpi.java [hunk 4]
```diff
@@ -359,7 +364,7 @@
     private Marshaller marsh;
 
     /** Statistics. */
-    protected final TcpDiscoveryStatistics stats = new TcpDiscoveryStatistics();
+    protected TcpDiscoveryStatistics stats;
```
Comment (NSAmelchev) [inline modules/core/src/main/java/org/apache/ignite/spi/discovery/tcp/TcpDiscoverySpi.java:367]: Now we can get NPE from public methods until SPI was started.

File: modules/core/src/main/java/org/apache/ignite/spi/discovery/tcp/TcpDiscoverySpi.java [hunk 5]
```diff
@@ -2115,6 +2109,23 @@ protected void onExchange(DiscoveryDataPacket dataPacket, ClassLoader clsLdr) {
     @Override public void spiStart(@Nullable String igniteInstanceName) throws IgniteSpiException {
         initializeImpl();
 
+        MetricRegistry discoReg =
```
Comment (NSAmelchev) [inline modules/core/src/main/java/org/apache/ignite/spi/discovery/tcp/TcpDiscoverySpi.java:2112]: Can be one-line

File: modules/core/src/test/java/org/apache/ignite/spi/discovery/tcp/TcpDiscoverySpiMBeanTest.java [hunk 6]
```diff
@@ -64,21 +70,121 @@
...
+
+                if (grid.localNode().id().equals(crd.id())) {
+                    assertTrue(bean.getCoordinatorSinceTimestamp() > 0);
+                    assertTrue(discoReg.<LongMetric>findMetric("CoordinatorSince").value() > 0);
+                }
+
+                assertEquals(1L, bean.getNodesLeft());
+                assertEquals(1, discoReg.<IntMetric>findMetric("LeftNodes").value());
             }
+
```
Comment (NSAmelchev) [inline modules/core/src/test/java/org/apache/ignite/spi/discovery/tcp/TcpDiscoverySpiMBeanTest.java:187]: Why this changed?

File: modules/core/src/main/java/org/apache/ignite/spi/discovery/tcp/TcpDiscoverySpi.java (hunk 3 above)
Comment (nizhikov) [reply modules/core/src/main/java/org/apache/ignite/spi/discovery/tcp/TcpDiscoverySpi.java:1366]: fixed.

**Pull request merged**